modules/jupyternotebook/main.nf

```diff
@@ -0,0 +1,98 @@
+
+    //NB: You likely want to override this with a container containing all required
+    //dependencies for you analysis. The container at least needs to contain the
+    //yaml and rmarkdown R packages.
+    //TODO: what container to use as default image?
+    conda (params.enable_conda ? "ipykernel=6.0.3 jupytext=1.11.4 nbconvert=6.1.0 papermill=2.3.3 matplotlib=3.4.2" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/YOUR-TOOL-HERE"
+    } else {
+        container "quay.io/biocontainers/YOUR-TOOL-HERE"
+    }
+
+    input:
+    tuple val(meta), path(notebook)
+    val(parameters)
+    path(input_files)
+
+    output:
+    tuple val(meta), path("*.html"), emit: report
+    path("artifacts/*"), emit: artifacts, optional: true
+    path "*.version.txt"          , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    // Dump parameters to yaml file.
+    // Using a yaml file over using the CLI params because
+    //  * no issue with escaping
+    //  * allows to pass nested maps instead of just single values
+    def params_cmd = ""
+    def render_cmd = ""
+    if (params.parametrize) {
+        nb_params = [:]
+        if (params.implicit_params) {
+            nb_params["cpus"] = task.cpus
+            nb_params["artifact_dir"] = "artifacts"
+            nb_params["input_dir"] = "./"
+        }
+        if (params.meta_params) {
+            nb_params["meta"] = meta
+        }
+        nb_params += parameters
+        params_cmd = dump_params_yml(nb_params)
+        render_cmd = "papermill -f .params.yml"
+    } else {
+        render_cmd = "papermill"
+    }
+
+    """
+    # Create output directory
+    mkdir artifacts
+
+    # Set parallelism for BLAS/MKL etc. to avoid over-booking of resources
+    export MKL_NUM_THREADS="${task.cpus}"
+    export OPENBLAS_NUM_THREADS="${task.cpus}"
+    export OMP_NUM_THREADS="${task.cpus}"
+    export NUMBA_NUM_THREADS="${task.cpus}"
+
+    # dump parameters to yaml
+    ${params_cmd}
+
+    # Convert notebook to ipynb using jupytext, execute using papermill, convert using nbconvert
+    jupytext --to notebook --output - --set-kernel - ${notebook}  \\
+        | ${render_cmd} \\
+        | jupyter nbconvert --stdin --to html --output ${prefix}.html
+
+    # TODO how to output versions of multiple tools?
+    echo ${software},jupytext,\$(jupytext --version) > versions.csv
+    echo ${software},ipykernel,\$(python -c "import ipykernel; print(ipykernel.__version__)") >> versions.csv
+    echo ${software},nbconvert,\$(jupyter nbconvert --version) >> versions.csv
+    echo ${software},papermill,\$(papermill --version | cut -f1 -d' ') >> versions.csv
```

Comment:
```suggestion
    cat <<-END_VERSIONS > versions.csv 
    ${software},jupytext,\$(jupytext --version)
    ${software},ipykernel,\$(python -c "import ipykernel; print(ipykernel.__version__)")
    ${software},nbconvert,\$(jupyter nbconvert --version)
    ${software},papermill,\$(papermill --version | cut -f1 -d' ')
    END_VERSIONS
```
I think using `cat` here is less prone to errors than using multiple `echo`'s and redirects

Reply:
I agree, but I will argue that the version of @grst is more newbie-friendly.

Reply:
I think using a heredoc is still beginner-friendly, as long as it's explained in the comment text of the template module. I agree that a beginner would be more familiar with `echo` and likely never heard of a heredoc, but I do not think that should stop us from using it if it results in code which is easier to read and debug. I think a mistyped `>` which should be a `>>` for example is easy to overlook. A beginner would still need to figure out the version number wrangling, which I think can be potentially more complex than understanding a heredoc allows for formatting text directly as input.

Reply:
Fair enough 👍

Reply:
I haven't heard of heredocs, but I think the above looks relatively straightforward to work out

I would recommend adding some documentation in the nf-core/modules guidelines, and pasting this as an example.

Reply:
Here's the related PR by @mahesh-panchal: https://github.com/nf-core/tools/pull/1247